rust/kernel/error.rs

```diff
@@ -104,3 +104,55 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+// # Invariant: `-bindings::MAX_ERRNO` fits in an `i16`.
+crate::static_assert!(bindings::MAX_ERRNO as i64 <= -(i16::MIN as i64));
+
+/// Transform a kernel "error pointer" to a normal pointer.
+///
+/// Some kernel C API functions return an "error pointer" which optionally
+/// embeds an `errno`. Callers are supposed to check the returned pointer
+/// for errors. This function performs the check and converts the "error pointer"
+/// to a normal pointer in an idiomatic fashion.
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// fn devm_platform_ioremap_resource(
+///     pdev: &mut PlatformDevice,
+///     index: u32,
+/// ) -> Result<*mut c_types::c_void> {
+///     // SAFETY: FFI call.
+///     unsafe {
+///         ptr_err_check(bindings::devm_platform_ioremap_resource(
+///             pdev.to_ptr(),
+///             index,
+///         ))
+///     }
+/// }
+/// ```
+// TODO: remove `dead_code` marker once an in-kernel client is available.
+#[allow(dead_code)]
+pub(crate) unsafe fn ptr_err_check<T>(ptr: *mut T) -> Result<*mut T> {
```

Comment:
Why name it as `ptr_err_check`? `check` sounds like an operation to me, and a `*_check` function seems like a function will return nothing or `bool`. Just want to know the reason of the name ;-)

Maybe we can name it as `ptr_or_err` or something?

Reply:
Also I don't think we need the `unsafe` for this function: we don't dereference the pointer and we know there is no way this function will cause UB.

Reply:
We can give it any name. Let's think of one.

Good point, it doesn't have to be an unsafe function. Move the two calls to unsafe functions in an `unsafe` block?

Reply:
The "official" name of this kernel abstraction is "error pointer". So how about `from_kernel_error_pointer()` ?

Reply:
> Good point, it doesn't have to be an unsafe function. Move the two calls to unsafe functions in an `unsafe` block?

I think we are under the rule that one `unsafe` block could only contain one unsafe function, so you have to use two `unsafe` blocks...